### PR TITLE
data submission

### PR DESCRIPTION
**Pull request by leoraba**
- new POST /submission/category/{categoryId} to upload submission
- parse TSV files into Schema entities
- Save or Update an Active Submission
- Swagger files

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,21 +16,32 @@ This projects uses:
 
 For quick start development use `docker-compose up -d` to start postgress database
 
-For database migration use `pnpm --filter common migrate`
+For database migration:
+
+- Set DB_URL environment variable `export DB_URL=[YOUR_DATABASE_CONNECTION_URL]`
+- Run the command `pnpm --filter common migrate`
```
Comment (leoraba) [inline README.md:22]: this environment variables is required for the migration script.

File: README.md [hunk 2]
```diff
@@ -16,21 +16,32 @@ This projects uses:
...
+| -------------- | ---------------------------- | ------- |
+| `PORT`         | Server Port.                 | 3030    |
+| `UPLOAD_LIMIT` | Limit upload file size       | '50mb'  |
+| `DB_HOST`      | Database Hostname            |         |
+| `DB_PORT`      | Database Port                |         |
+| `DB_NAME`      | Database Name                |         |
+| `DB_USER`      | Database User                |         |
+| `DB_PASSWORD`  | Database Password            |         |
+| `LECTERN_URL`  | Schema Service (Lectern) URL |         |
+| `LOG_LEVEL`    | Log Level                    | 'info'  |
```
Comment (leoraba) [inline README.md:44]: Adding all used environment variables on README file

File: apps/server/src/config/swagger.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ const swaggerDefinition = {
 const options = {
 	swaggerDefinition,
 	// Paths to files containing OpenAPI definitions
-	apis: ['./src/routes/*.ts'],
+	apis: ['./src/routes/*.ts', './src/routes/swagger/*.yml'],
```
Comment (leoraba) [inline apps/server/src/config/swagger.ts:17]: this generates the swagger output file based on swagger annotations and YML files.

File: apps/server/src/routes/swagger/dictionary-api.yml
Comment (leoraba) [inline apps/server/src/routes/swagger/dictionary-api.yml:1]: Moved all swagger annotations from `package/common/..` folders in to the `apps/server/...` folder. So now the server is responsible to generate swagger output file

File: packages/common/migrations/migrate.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
-import { drizzle } from 'drizzle-orm/postgres-js';
-import { migrate } from 'drizzle-orm/postgres-js/migrator';
-import { Pool } from 'pg';
+import { drizzle } from 'drizzle-orm/node-postgres';
+import { migrate } from 'drizzle-orm/node-postgres/migrator';
```
Comment (leoraba) [inline packages/common/migrations/migrate.ts:2]: Changed to `node-postgres` to be able to use connection Pools

File: packages/common/package.json [hunk 5]
```diff
@@ -19,16 +19,20 @@
 	"type": "module",
 	"dependencies": {
 		"@overturebio-stack/lectern-client": "^1.5.0",
+		"deep-freeze": "^0.0.1",
 		"dotenv": "^16.4.1",
 		"drizzle-orm": "^0.29.3",
 		"express": "^4.18.2",
 		"lodash-es": "^4.17.21",
+		"multer": "1.4.5-lts.1",
```
Comment (leoraba) [inline packages/common/package.json:27]: to handle file uploads

File: packages/common/src/config/config.ts [hunk 6]
```diff
@@ -32,7 +33,7 @@ export type AppConfig = {
  * Dependencies required for utils/services
  */
 export interface Dependencies {
-	db: NodePgDatabase;
+	db: NodePgDatabase<typeof schema>;
```
Comment (leoraba) [inline packages/common/src/config/config.ts:36]: Required by drizzle to query nested relational data

File: packages/common/src/controllers/dictionaryController.ts [hunk 7]
```diff
@@ -12,7 +12,7 @@ const controller = (dependencies: Dependencies) => {
 		getCurrentDictionary: async (req: Request, res: Response, next: NextFunction) => {
 			try {
 				//TODO: implement logic to get current dictionary from DB
-				throw new NotImplemented('This functionallity is not implemented');
+				throw new NotImplemented();
```
Comment (leoraba) [inline packages/common/src/controllers/dictionaryController.ts:15]: `NotImplemented` class sets this message by default on it's empty constructor

File: packages/common/src/controllers/submissionController.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+				logger.info(
+					LOG_MODULE,
+					`Upload Submission Request categoryId '${categoryId}' files '${files.map((f) => f.originalname)}'`,
+				);
+
+				const fileErrors: BatchError[] = [];
+				const rawSubmissionEntities: SubmissionEntity[] = [];
+
+				for (const file of files) {
+					try {
```
Comment (leoraba) [inline packages/common/src/controllers/submissionController.ts:28]: Iterating the files that user sends us, and perform a tsv format validations (valid .tsv extension, parsing tsv to json, ..)

File: packages/common/src/controllers/submissionController.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+							message: getErrorMessage(error),
+							type: BATCH_ERROR_TYPE.TSV_PARSING_FAILED,
+							batchName: file.originalname,
+						};
+
+						fileErrors.push(batchError);
+					}
+				}
+
+				const resultSubmission = await service.uploadSubmission(rawSubmissionEntities, categoryId);
```
Comment (leoraba) [inline packages/common/src/controllers/submissionController.ts:54]: passing to the service all submissions to be validated.

File: packages/common/src/repository/activeSubmissionRepository.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import { SelectedFields } from 'drizzle-orm/pg-core/query-builders/select.types';
+import { SQL } from 'drizzle-orm/sql';
+import { isEmpty } from 'lodash-es';
+
+import { Dependencies } from '../config/config.js';
+import { NewSubmission, Submission, submissions } from '../models/submissions.js';
+import { ServiceUnavailable } from '../utils/errors.js';
+
+const repository = (dependencies: Dependencies) => {
```
Comment (leoraba) [inline packages/common/src/repository/activeSubmissionRepository.ts:9]: Repositories to separate database queries

File: packages/common/src/routers/submissionRouter.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+	router.use(urlencoded({ extended: false }));
+	router.use(json());
+
+	router.get('/category/:categoryId', auth, submissionControllers(dependencies).listActive);
+
+	router.post('/category/:categoryId/upload', upload.array('files'), submissionControllers(dependencies).upload);
+
+	router.post('/category/:categoryId/commit/:id', auth, submissionControllers(dependencies).commit);
+	return router;
+};
```
Comment (leoraba) [inline packages/common/src/routers/submissionRouter.ts:22]: Creating 3 routers for Submission. Although only `upload` route has implementation. The others return error 501.

File: packages/common/src/services/submissionService.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+			const { getCurrentDictionary } = dictionaryUtils(dependencies);
+
+			let resultSubmission: CreateActiveSubmission = emptySubmission();
+			let batchErrors: BatchError[] = [];
+
+			if (submissionsEntities.length > 0) {
+				const currentDictionary = await getCurrentDictionary(categoryId);
+				// TODO: map this dictionary field as a 'DictionaryData' type on Drizzle models
+				// Dictionary data stores as a JSONB type in database
+				const dictionaryData: DictionaryData[] = currentDictionary.dictionary as DictionaryData[];
```
Comment (leoraba) [inline packages/common/src/services/submissionService.ts:41]: so far not running any DictionarySchemas to validate. Will require use corresponding `SchemaDefinition` type to run validations

File: packages/common/src/utils/fileUtils.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+ * @param {string} fileName
+ * @returns a JSON format objet
+ */
+export const tsvToJson = async (fileName: string): Promise<ReadonlyArray<TsvRecordAsJsonObj>> => {
+	const contents = await fsPromises.readFile(fileName, 'utf-8');
+	const arr = parseTsvToJson(contents);
+	return arr;
+};
+
+const parseTsvToJson = (content: string): ReadonlyArray<TsvRecordAsJsonObj> => {
```
Comment (leoraba) [inline packages/common/src/utils/fileUtils.ts:34]: After some failed attempts to find a most robust library to parse TSVs to JSON, found more convenient to reuse most of this logic from argo-clinical.

File: packages/common/src/utils/submissionUtils.ts [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+			const foundOpenSubmission = await utils(dependencies).getCurrentActiveSubmission(Number(categoryId));
+			let updatedSubmission: Submission;
+			if (!isEmpty(foundOpenSubmission)) {
+				const { id, data } = foundOpenSubmission[0];
+				const currentSubmissionId = Number(id);
+				const currentData = data as Record<string, SubmissionEntity>;
+				logger.debug(LOG_MODULE, `Found an Active submission`, JSON.stringify(currentData));
+
+				// merge current active submission data
+				const newData = { ...currentData, ...entityMap };
```
Comment (leoraba) [inline packages/common/src/utils/submissionUtils.ts:34]: Appends or update current submission data

File: apps/server/src/server.ts [hunk 15]
```diff
@@ -18,6 +18,7 @@ app.use(helmet());
 
 app.use('/ping', pingRouter);
 app.use('/dictionary', lyricProvider.routers.dictionary);
+app.use('/submission', lyricProvider.routers.submssion);
```
Comment (justincorrigible) [inline apps/server/src/server.ts:21]: local typo?
```suggestion
app.use('/submission', lyricProvider.routers.submission);
```

File: packages/common/migrations/0001_abnormal_gressill.sql [hunk 16]
```diff
@@ -0,0 +1,5 @@
+ALTER TYPE "submission_state" ADD VALUE 'OPEN';--> statement-breakpoint
+ALTER TYPE "submission_state" ADD VALUE 'VALID';--> statement-breakpoint
+ALTER TYPE "submission_state" ADD VALUE 'INVALID';--> statement-breakpoint
+ALTER TYPE "submission_state" ADD VALUE 'CLOSED';--> statement-breakpoint
+ALTER TYPE "submission_state" ADD VALUE 'COMMITED';
```
Comment (justincorrigible) [inline packages/common/migrations/0001_abnormal_gressill.sql:5]: may also need to check the spelling elsewhere
```suggestion
ALTER TYPE "submission_state" ADD VALUE 'COMMITTED';
```

File: packages/common/src/core/provider.ts [hunk 17]
```diff
@@ -22,6 +22,7 @@ const provider = (configData: AppConfig) => {
 		configs: deps,
 		routers: {
 			dictionary: dictionaryRouters(deps),
+			submssion: submissionRouters(deps),
```
Comment (justincorrigible) [inline packages/common/src/core/provider.ts:25]: found it!
```suggestion
			submission: submissionRouters(deps),
```

File: packages/common/src/models/submissions.ts [hunk 18]
```diff
@@ -5,7 +5,7 @@ import { integer, jsonb, pgEnum, pgTable, serial, timestamp, varchar } from 'dri
 import { dictionaries } from './dictionaries.js';
 import { dictionaryCategories } from './dictionary_categories.js';
 
-export const submissionStateEnum = pgEnum('submission_state', ['open', 'valid', 'invalid']);
+export const submissionStateEnum = pgEnum('submission_state', ['OPEN', 'VALID', 'INVALID', 'CLOSED', 'COMMITED']);
```
Comment (justincorrigible) [inline packages/common/src/models/submissions.ts:8]: found it!
```suggestion
export const submissionStateEnum = pgEnum('submission_state', ['OPEN', 'VALID', 'INVALID', 'CLOSED', 'COMMITTED']);
```

File: packages/common/src/routers/submissionRouter.ts (hunk 11 above)
Comment (justincorrigible) [reply packages/common/src/routers/submissionRouter.ts:22]: what will be the difference between uploading and committing?

File: packages/common/src/utils/categoryUtils.ts [hunk 19]
```diff
@@ -33,6 +34,22 @@ const utils = (dependencies: Dependencies) => {
 				throw error;
 			}
 		},
+
+		/**
+		 * Finds a Category instance based on its unique ID
+		 * @param {niumber} categoryId The ID of the Category
```
Comment (justincorrigible) [inline packages/common/src/utils/categoryUtils.ts:40]:
```suggestion
		 * @param {number} categoryId The ID of the Category
```

File: packages/common/src/routers/submissionRouter.ts (hunk 11 above)
Comment (leoraba) [reply packages/common/src/routers/submissionRouter.ts:22]: `Upload` puts data in submission for validation and allows to fix errors, add multiple entities, etc. . `Commit` is an user confirmation that the submission is completed.

**Comment by leoraba** [comment]
Typos fixed and committed.
Merging this PR(to avoid conflicts) as it's a dependency of PR#18

**Pull request merged**